R/module_init_data.R

```diff
@@ -69,55 +69,29 @@ srv_init_data <- function(id, data, modules, filter = teal_slices()) {
 
     # data_rv contains teal_data object
     # either passed to teal::init or returned from teal_data_module
+
     data_validated <- if (inherits(data, "teal_data_module")) {
+      shinyjs::disable(selector = sprintf(".teal-body:has('#%s') .nav li a", session$ns("content")))
       srv_teal_data(
         "teal_data_module",
-        data = reactive(req(FALSE)), # to .fallback_on_failure to shiny.silent.error
+        data = reactive(teal_data()),
         data_module = data,
         modules = modules,
-        validate_shiny_silent_error = FALSE
+        validate_shiny_silent_error = FALSE,
+        failure_callback = function(data) {
+          if (inherits(data(), "teal_data")) {
+            shinyjs::enable(selector = sprintf(".teal-body:has('#%s') .nav li a", session$ns("content")))
+          } else {
+            shinyjs::disable(selector = sprintf(".teal-body:has('#%s') .nav li a", session$ns("content")))
+          }
+        }
       )
     } else if (inherits(data, "teal_data")) {
       reactiveVal(data)
     } else if (test_reactive(data)) {
-      .fallback_on_failure(this = data, that = reactive(req(FALSE)), label = "Reactive data")
-    }
-
-    if (inherits(data, "teal_data_module")) {
-      shinyjs::disable(selector = sprintf(".teal-body:has('#%s') .nav li a", session$ns("content")))
+      data
     }
 
-    observeEvent(data_validated(), {
-      showNotification("Data loaded successfully.", duration = 5)
-      shinyjs::enable(selector = sprintf(".teal-body:has('#%s') .nav li a", session$ns("content")))
-      if (isTRUE(attr(data, "once"))) {
-        # Hiding the data module tab.
-        shinyjs::hide(
-          selector = sprintf(
-            ".teal-body:has('#%s') a[data-value='teal_data_module']",
-            session$ns("content")
-          )
-        )
-        # Clicking the second tab, which is the first module.
-        shinyjs::runjs(
-          sprintf(
-            "document.querySelector('.teal-body:has(#%s) .nav li:nth-child(2) a').click();",
-            session$ns("content")
-          )
-        )
-      }
-
-      is_filter_ok <- check_filter_datanames(filter, .teal_data_ls(data_validated()))
-      if (!isTRUE(is_filter_ok)) {
-        showNotification(
-          "Some filters were not applied because of incompatibility with data. Contact app developer.",
-          type = "warning",
-          duration = 10
-        )
-        warning(is_filter_ok)
-      }
-    })
```

Comment:
This removal ended up in this fail of the test

```r
  `actual`:   TRUE 
  `expected`: FALSE
  ── Failure ('test-shinytest2-teal_data_module.R:167:3'): e2e: teal_data_module gets removed after successful data load, when once = TRUE ──
  app$is_visible(sprintf("#%s", submit)) is not FALSE
```

Should we bring this code back so that `teal_data_module` is hidden for `once` set to `TRUE`?

Reply:
Yes, but only the part responsible for `once`

Reply:
2 tests are failing for this

```r
Failure (test-shinytest2-teal_data_module.R:163:3): e2e: teal_data_module gets removed after successful data load, when once = TRUE
app$is_visible("#teal-teal_modules-active_tab a[data-value=\"teal_data_module\"]") is not FALSE

`actual`:   TRUE 
`expected`: FALSE

Failure (test-shinytest2-teal_data_module.R:167:3): e2e: teal_data_module gets removed after successful data load, when once = TRUE
app$is_visible(sprintf("#%s", submit)) is not FALSE

`actual`:   TRUE 
`expected`: FALSE


```

Reply:
https://github.com/insightsengineering/teal/pull/1341/commits/8265176a4f6376278cb4ab87b10a58f5bdc4754d